beetsplug/fetchart.py

```diff
@@ -122,22 +128,49 @@ def _validate(self, plugin):
                                 self.size[0], self.size[1])
                 return self.CANDIDATE_BAD
 
-        # Check maximum size.
+        # Check maximum dimension.
+        downscale = False
         if plugin.maxwidth and self.size[0] > plugin.maxwidth:
-            self._log.debug(u'image needs resizing ({} > {})',
+            self._log.debug(u'image needs rescaling ({} > {})',
                             self.size[0], plugin.maxwidth)
+            downscale = True
+
+        # Check filesize.
+        filesize = os.stat(syspath(self.path)).st_size
+        downsize = False
+        if plugin.max_filesize and filesize > plugin.max_filesize:
+            self._log.debug(u'image needs resizing ({}B > {}B)',
+                            filesize, plugin.max_filesize)
+            downsize = True
+
+        if downscale and downsize:
+            return self.CANDIDATE_DOWNSCALE_AND_DOWNSIZE
+        elif downscale and not downsize:
             return self.CANDIDATE_DOWNSCALE
-
-        return self.CANDIDATE_EXACT
+        elif downsize and not downscale:
+            return self.CANDIDATE_DOWNSIZE
+        else:
+            return self.CANDIDATE_EXACT
 
     def validate(self, plugin):
         self.check = self._validate(plugin)
         return self.check
 
     def resize(self, plugin):
-        if plugin.maxwidth and self.check == self.CANDIDATE_DOWNSCALE:
+        if self.check == self.CANDIDATE_DOWNSCALE_AND_DOWNSIZE:
+            self.path = \
+                ArtResizer.shared.resize(plugin.maxwidth, self.path,
+                                         quality=plugin.quality,
+                                         max_filesize=plugin.max_filesize)
+        elif self.check == self.CANDIDATE_DOWNSCALE:
             self.path = ArtResizer.shared.resize(plugin.maxwidth, self.path,
                                                  quality=plugin.quality)
+        elif self.check == self.CANDIDATE_DOWNSIZE:
+            # dimensions are correct, so maxwidth is set to maximum dimension
+            self.path = \
+                ArtResizer.shared.resize(max(self.size), self.path,
+                                         quality=plugin.quality,
+                                         max_filesize=plugin.max_filesize)
```

Comment:
Wouldn't the following have the same effect? I don't see the need to set the `maxwidth` or `max_filesize` arguments in this way, the `resize` method should behave (essentially) the same
```suggestion
        if self.check in [self.CANDIDATE_DOWNSCALE,
                self.CANDIDATE_DOWNSIZE,
                self.CANDIDATE_DOWNSCALE_AND_DOWNSIZE]:
            self.path = ArtResizer.shared.resize(plugin.maxwidth, self.path,
                                         quality=plugin.quality,
                                         max_filesize=plugin.max_filesize)
```
Wouldn't that also mean that just one flag `CANDIDATE_DOWNSIZE` is enough since the action is the same for all three cases?

Reply:
Your suggestion is certainly neater, however there remains a use-case where a user may want to target a maximum pixel width but doesn't care about the filesize, thus would prefer the original functionality of `CANDIDATE_DOWNSCALE`. In this case, `max_filesize` may have not been set, or the file is already lower than the maximum filesize.

Thanks for pointing out that `CANDIDATE_DOWNSIZE` is exactly the same as `CANDIDATE_DOWNSCALE_AND_DOWNSIZE` - I'll condense these down into one option.

Reply:
Yeah sorry for deleting my earlier comment, I was trying to think through all of the possible combinations & any downsides. I think a good compromise, due to the default behaviour as you've pointed out, is to combine the two `DOWNSCALE` options. The `DOWNSIZE`-only option still has a use for those who don't want to rescale their images, just lower filesize (a use-case relevant for the Twitter example).

The only good reason I can think of to keep the combined option would be if setting `max_filesize` incurred a performance penalty when using the ImageMagick backend with an image _already below_ that filesize, however that's extremely unlikely. Using the PIL backend it would have no effect. Initially I wrote the code like this because I wasn't sure we could trust the initial values (I misinterpreted the checks for `if plugin.maxwidth` as them being set to `None` - I realise now they are initialised to zero in the default config). Thanks so much for your scrutiny here! Apologies again for my confusion.